src/game/action/attack.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright � 2020 uwadmin12
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "objectaction.h"
+#include "../object/creature.h"
+
+namespace reone {
+
+namespace game {
+
+class AttackAction : public ObjectAction {
+public:
+    AttackAction(const std::shared_ptr<Creature> &object, bool pursue, float distance = 3.2f, float timeout = 6.0f) :
+        ObjectAction(ActionType::AttackAction, object), _timeout(timeout), _inPursuit(pursue), _distance(distance) { }
+
+    std::shared_ptr<Creature> target() { return std::static_pointer_cast<Creature>(_object); } // static_pointer_cast(nullptr)?
+
+    void setAttack() { _inPursuit = false;  }
+
+    /* not in pursuit means attack signal is high, be it ranged or melee */
+    bool isInPursuit() { return _object != nullptr && _inPursuit; }
+
+    float& getTimeout() { return _timeout; }
```

Comment:
This breaks encapsulation and is a little counter-intuitive 🙂 In OOP, objects should receive commands and act on their (hidden) internal state. How about this?

```cpp
void advance(float dt) {
    _timeout = glm::max(0.0f, _timeout - dt);
}

bool isTimedOut() const {
    return _timeout == 0.0f;
}
```